### PR TITLE
[SDK-1837] Session storage support for transactions

### LINKED ISSUE #319: Cookie Pollution - still happening to us with latest version
Hello,

We are using auth0-spa-js in our product (perception point X-ray) 

And are still encountering the cookie pollution issue reported in:
https://github.com/auth0/auth0-spa-js/issues/217

Even after upgrading auth0-spa-js to latest version and deploying to production every week.

### Description
Sometimes our users encounter this error when trying to log in to our web app.
It's hard to determine when and why this happens.

image

Until the user will not manually clear his cookies for our domain in his browser, he will not be able to log in.

Here is a screenshot of an example for all the cookies stored for our domain on the user's browser when the error was received:

image

Example cookie:
domain: `xray.testing.perception-point.io`
name:`a0.spajs.txs.LmtkdVhkb3I4aEhBflRZY3dqVG1UU1pVcn5qYmF6MVRfS0RuVXBQNWtRTQ%3D%3D`
value: `{%22nonce%22:%22s0xI3.u6P30RgR0_H6APyOAS-KOoKh111q1.xrcRCJT%22%2C%22code_verifier%22:%22Ch6GBCJ-naY7pFWa8Mq19yjJ35F0LSwKGzlYwR6wQCG%22%2C%22appState%22:{%22returnPath%22:%22/scans?dateTimeRange=preset-lastDay&freeText=liron&origin=VOID&verdict=VOID%22}%2C%22scope%22:%22openid%20profile%20email%20user_metadata%20app_metadata%20picture%22%2C%22audience%22:%22https://xray.perception-point.io/%22}`
expires: `2019-12-17T09:35:16.000Z`

### Environment

Domain:
https://xray.testing.perception-point.io/

OS:
Windows 10

Browser:
Google Chrome Version 79.0.3945.79

Library version:
1.5.0

### LINKED ISSUE #515: Invalid State Error on Ipad  safari and chrome
### Describe the problem

> When logging in with a **loginWithRedirect**, you will be redirected back to your app with code and state values set as URL query parameters. These values are used by **handleRedirectCallback** And this result in Error: Invalid state

 
### Environment

> Please provide the following:

- **Version of `auth0-spa-js` used:** 1.10.0
- **Which browsers have you tested in?** Ipad "Chrome Mobile iOS 83.0.4103 / iOS 12.0.0", "Mobile Safari 12.0.0 / iOS 12.0.0", 
- **Which framework are you using, if applicable (Angular, React, etc):** "React"

### PR DESCRIPTION
**Pull request by stevehobbsdev**
### Description

This PR modifies the internal transaction mechanism to use session storage instead of cookies. Furthermore, only one transaction can be present at a time.

Motivation is to fix potentially issues where the transaction is not cleaned up after auth, leading to a build up of transaction cookies that could break users.

Also fixes an issue with transaction cookies not working with Safari/macOS.

### References

Ref #515 
Fixes #319 

### Testing

Added unit tests, but also tested manually across latest Chrome, Firefox, Safari, Edge and IE11.

- [X] This change adds test coverage for new/changed/fixed functionality

### Checklist

- [X] I have added documentation for new/changed functionality in this PR or in auth0.com/docs
- [X] All active GitHub checks for tests, formatting, and security are passing
- [X] The correct base branch is being used, if not `master`

### CODE REVIEW

File: src/storage.ts [hunk 1]
```diff
@@ -4,30 +4,61 @@ interface ClientStorageOptions {
...
+    if (typeof value === 'undefined') {
+      return;
+    }
+    return <T>JSON.parse(value);
+  },
+
+  save(key: string, value: any, options?: ClientStorageOptions): void {
+    let cookieAttributes: Cookies.CookieAttributes = {};
+    if ('https:' === window.location.protocol) {
+      cookieAttributes = {
```
Comment (adamjmcgrath) [inline src/storage.ts:31]: People affected by #515 will still have an issue with the `is.authenticated` cookie

We'll need to do the [2 cookie solution](https://web.dev/samesite-cookie-recipes/#handling-incompatible-clients) on that before we can say it's fixed

File: __tests__/index.test.ts [hunk 2]
```diff
@@ -4,8 +4,21 @@ import * as scope from '../src/scope';
 // @ts-ignore
 import { acquireLockSpy, releaseLockSpy } from 'browser-tabs-lock';
 
+jest.mock('browser-tabs-lock');
```
Comment (adamjmcgrath) [inline __tests__/index.test.ts:7]: not needed - node modules are automatically mocked

File: __tests__/index.test.ts [hunk 3]
```diff
@@ -51,6 +64,13 @@ const mockEnclosedCache = {
   clear: jest.fn()
 };
 
+const mockStorage = {
```
Comment (adamjmcgrath) [inline __tests__/index.test.ts:67]: unused

File: __tests__/index.test.ts [hunk 4]
```diff
@@ -64,6 +84,15 @@ const webWorkerMatcher = expect.objectContaining({
   postMessage: expect.any(Function)
 });
 
+const transaction = {
```
Comment (adamjmcgrath) [inline __tests__/index.test.ts:87]: unused

File: __tests__/index.test.ts [hunk 5]
```diff
@@ -73,6 +102,7 @@ const setup = async (clientOptions: Partial<Auth0ClientOptions> = {}) => {
     remove: require('../src/storage').remove
   };
 
+  const lock = require('browser-tabs-lock');
```
Comment (adamjmcgrath) [inline __tests__/index.test.ts:105]: unused

File: src/Auth0Client.ts [hunk 6]
```diff
@@ -748,9 +744,10 @@ export default class Auth0Client {
       options.timeoutInSeconds || this.options.authorizeTimeoutInSeconds;
     const codeResult = await runIframe(url, this.domainUrl, timeout);
 
-    if (stateIn !== codeResult.state) {
```
Comment (adamjmcgrath) [inline src/Auth0Client.ts:751]: Why do you need to remove these checks?

**Review by adamjmcgrath — CHANGES_REQUESTED** [review]

File: __tests__/index.test.ts (hunk 2 above)
Comment (stevehobbsdev) [reply __tests__/index.test.ts:7]: Not sure where that has come from - I had a conflict here, possibly a dodgy merge.

File: src/Auth0Client.ts [hunk 7]
```diff
@@ -748,9 +744,10 @@ export default class Auth0Client {
...
-      throw new Error('Invalid state');
-    }
+    // if (stateIn !== codeResult.state) {
```
Comment (trevorgk) [inline src/Auth0Client.ts:747]: commented code

File: src/storage.ts (hunk 1 above)
Comment (stevehobbsdev) [reply src/storage.ts:31]: They should no longer get the invalid state issue though, which is what he original issue reported.

File: src/Auth0Client.ts (hunk 6 above)
Comment (stevehobbsdev) [reply src/Auth0Client.ts:751]: That was an error on my part while I was investigating something, the intention was to put them back in.

File: src/storage.ts (hunk 1 above)
Comment (adamjmcgrath) [reply src/storage.ts:31]: true, we can handle that in another PR - but we should include it in the next release

i've started a branch here https://github.com/auth0/auth0-spa-js/tree/legacy-samesite

File: __tests__/index.test.ts (hunk 2 above)
Comment (adamjmcgrath) [inline __tests__/index.test.ts:7]:
```suggestion
```

File: __tests__/index.test.ts [hunk 8]
```diff
@@ -419,18 +424,6 @@ describe('Auth0', () => {
       expectToHaveBeenCalledWithAuth0ClientParam(utils.runPopup, auth0Client);
     });
 
-    it('throws error if state from popup response is different from the provided state', async () => {
```
Comment (adamjmcgrath) [inline __tests__/index.test.ts:422]: You can put this test back

File: __tests__/index.test.ts [hunk 9]
```diff
@@ -1897,24 +1887,6 @@ describe('Auth0', () => {
         );
       });
 
-      it('throws error if state from popup response is different from the provided state', async () => {
```
Comment (adamjmcgrath) [inline __tests__/index.test.ts:1900]: You can put this test back

File: src/Auth0Client.ts [hunk 10]
```diff
@@ -753,6 +749,8 @@ export default class Auth0Client {
       throw new Error('Invalid state');
     }
 
+    if (!codeResult) throw 'Error';
```
Comment (adamjmcgrath) [inline src/Auth0Client.ts:752]: unneeded

**Review by adamjmcgrath — CHANGES_REQUESTED** [review]

**Comment by adamjmcgrath** [comment]
Integrating https://github.com/stevehobbsdev/auth0-spa-js/pull/1 should fix your flakey build

File: src/Auth0Client.ts [hunk 11]
```diff
@@ -337,10 +337,6 @@ export default class Auth0Client {
         DEFAULT_AUTHORIZE_TIMEOUT_IN_SECONDS
     });
 
-    if (stateIn !== codeResult.state) {
```
Comment (adamjmcgrath) [inline src/Auth0Client.ts:340]: This should go back too?

File: __tests__/index.test.ts (hunk 2 above)
Comment (stevehobbsdev) [reply __tests__/index.test.ts:7]: What does the hyphen prefix do here?

File: src/Auth0Client.ts (hunk 10 above)
Comment (stevehobbsdev) [reply src/Auth0Client.ts:752]: ಠ_ಠ not sure where that has come from!

File: __tests__/index.test.ts (hunk 2 above)
Comment (stevehobbsdev) [reply __tests__/index.test.ts:7]: Oh, I was looking at this diff on my phone, it wasn't obvious it was part of the diff markup 😄

Comment (stevehobbsdev) [reply __tests__/index.test.ts:7]: Have removed it manually.

File: __tests__/index.test.ts [hunk 12]
```diff
@@ -1897,24 +1899,6 @@ describe('Auth0', () => {
...
```
Comment (adamjmcgrath) [inline __tests__/index.test.ts:1900]: Just need to put this back, then we're good

**Review by adamjmcgrath — APPROVED** [review]

**Comment by ohararyan** [comment]
@stevehobbsdev do we know when this might be tagged for release?

**Comment by stevehobbsdev** [comment]
@ohararyan We're working to push a release out by the end of the week, I'll tag this thread once it's out 👍

**Comment by vibaiher-qatium** [comment]
Hi! We think we have detected an issue with this feature.

We are using @auth0/auth0-react using Magic Link configuration. If we click the link, the browser opens it in a new tab window where the session storage is not set, so transaction is not found and user is not authenticated. If we copy that link and paste in same tab, then the user is authenticated.

Following this: https://developer.mozilla.org/en-US/docs/Web/API/Window/sessionStorage looks like new tab windows does not share same session storage.

**Comment by stevehobbsdev** [comment]
@vibaiher-qatium I've created a new issue for this on your behalf so it does not get lost: #590
